agent/consul/stream/subscribe.proto

```diff
@@ -0,0 +1,51 @@
+syntax = "proto3";
+
+package stream;
+
+import "structs.proto";
+
+service Consul {
+    rpc Subscribe(SubscribeRequest) returns (stream Event) {}
+} 
+
+enum Topic {
+    ServiceHealth = 0;
+}
+  
+message SubscribeRequest {
+    Topic Topic = 1;
+    string Key = 2;
+    string Token = 3;
+}
+
+message Event {
+    Topic Topic = 1;
+    string Key = 2;
+    uint64 Index = 3;
+    repeated ACLRule RequiredACLs = 4;
+    oneof Payload {
+        ServiceHealthUpdate ServiceHealth = 5;
+        bool EndOfSnapshot = 6;
+    }
+}
+
+enum ACLResource {
+    ServiceACL = 0;
+    NodeACL = 1;
+}
+
+message ACLRule {
+    ACLResource Resource = 1;
+    string Segment = 2; 
+}
```

Comment:
@mkeeler do you have an opinion on this representation and/or idea of embedding ACL rules into events?

The tl/dr is:
 - We want to define in one central place which types of ACL rules need to be checked for each event type.
 - Since events are generated by protobuf we have to do that somewhere in code (acl.go in this file needs to be updated along side event definitions but still all in one place).
 - Encoding them on the wire has a really nice future benefit - we can un-ambiguously enforce ACL at the client without needing to worry about compatibility between client and server versions etc. letting new types slip through or different rules be applied. This means that in the future clients can optimize and de-duplicate cache entries from multiple differnt clients while still ensureing they filter results according to the ACL rules the server would have applied.

This seems like a reasonable minimal represenation for those and it's internal only and can be evolved as we build more types and extend ACLs but interested if you see issues with this idea or encoding?

Reply:
If I understand this PR correctly, we will be enforcing ACLs server side so that we don't send events to clients that shouldn't see them but we want to also send the required rules to the client so that it can deliver the final results to all those http clients that are waiting for them and that have permissions to have the data.

If that is correct then I believe that this change is a good one however I don't believe it will be a silver bullet for future compatibility. If we add new ACL resources in the future the clients will still need to be updated to be made aware of those things. That is unless we rework the github.com/hashicorp/consul/acl package to handle resources/segments more generically. Then again the clients HTTP API wouldn't be able to handle those other types anyways so that probably doesn't matter. What it does buy us is the ability to change which available types of resources are required for enforcement. I could see this being especially important for config entry streaming eventually.

A couple other more general questions are:

1. Have we put any thought into how this would interact with Sentinel. Because the tokens/policies will get resolved both on the client and server the sentinel policy would execute on both. Streaming shouldn't affect it but it might be worth verifying.

2. Maybe I missed it in this PR (or potentially another one), but how do we handle multiple subscriptions using different tokens. Do we open up a new stream for the different token? Once streaming has started how does a client request subscriptions to additional topics?

Having not thought about the details nearly as much as either of you I would have expected something like:

1. open GRPC stream with some initial topic(s) and token(s)
2. server sends snapshot of the event data for the topic(s) filtering the events by the token(s)
3. server now streams events to the client
4. client requests subscription for new topic(s) with token(s)
5. server sends snapshot of the new data for the topic(s) filtering the events by the token(s)
6. Repeat 3-5 for a while
7. client unsubscribes from topic(s) and/or removes/updates token(s)
8. Back to 3-5 again
9. Eventually the stream gets closed

Basically with one stream there is a back and forth between the server and client to (un)subscribe to topic(s) and manage the ACL tokens used by the stream as many could be and probably will be involved.

Reply:
Great points Matt thanks for the sanity check.

> Have we put any thought into how this would interact with Sentinel. 

Nope. For now the ACL embedding is a future-proofing mechanism. I was actually only intending to build the mechanism to define those so this was possible later but I guess it doesn't hurt to actually transfer them already too. Sentinel would need to be executed on client too otherwise we'd need to special case ACLs that have sentinel attached and not allow that client side. This is less a comitment to doing things client side and more just an idea that makes the code here a little cleaner (all ACL needs in one place near type definitions) and also allows for future _possibility_ of more efficient clients with shared cache entries.

> Maybe I missed it in this PR (or potentially another one), but how do we handle multiple subscriptions using different tokens. Do we open up a new stream for the different token?

Yes. Agent cache is partitioned by Token implicitly so every client with a different token gets a separate cache entry for a given request currently. Obviously that might be wasteful when they all are the same and there are many clients/tokens on a single agent but that's not the common case and that is the simplest model for not leaking stuff. This addition was a step towards potentially being able to relax that in some cases later if we think that's worth the extra complexity.

> Once streaming has started how does a client request subscriptions to additional topics?

It doesn't. The current model is super simple - each `Subscribe` RPC is for a single topic. If a client needs multiple topics each one is a separate cache entry and a separate backend `Subscribe` call. In some cases that may mean the same event is sent redundantly on multiple subscriptions to the same client but the simplicity makes this a good trade - especial give the fact we are already several orders of magnitude better than before streaming.

We thought about more complex client behaviour where the client maintains just a single subscription and dynamically subscribes and unsubscribes to topics but this hugely complicates the client code and also snapshotting and playback semantics per topic on a shared connect etc. Unless the performance is a real-world issue I doubt we'd want to go there.

Reply:
This is very much a naïve question looking at this without context, but are we signing the contents of the Protobuf messages in some way to protect against something like a spoofed message embedding `required_acls: []`?

Reply:
Great question @mikemorris. We aren't because we rely on TLS for any secure communication.

This is a response from a server that we should be talking to over TLS and should have validated its identity on connection. If we don't to either of those then games already over for any MITM attack that wants to send us arbitrary stuff.